### PR TITLE
Server dedicated thread for routine processing [9488]

### PR DESCRIPTION
**Pull request by rsanchez15**
Rebase `feature/discovery-server/prototype` after https://github.com/eProsima/Fast-DDS/pull/1439 is merged.

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 1]
```diff
@@ -958,6 +958,11 @@ bool DiscoveryDataBase::delete_entity_of_change(
     return false;
 }
 
+bool DiscoveryDataBase::data_queue_empty()
+{
+    return data_queue_.Empty();
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:963]: I think this returns whether the Foreground queue is empty, where the Server pushes changes to the background queue. Foreground queue will always be empty after `process_data_queue()`. I think the easiest would be to check if both foreground and background are empty.
```suggestion
    return data_queue_.BothEmpty();
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.hpp [hunk 2]
```diff
@@ -123,11 +124,12 @@ class PDPServer2 : public aux::PDP
...
-    //! Wakes up the DServerEvent2 for new matching or trimming
-    void awakeServerThread()
-    {
-        mp_sync->restart_timer();
-    }
+    /*
+     * Wakes up the DServerEvent2 for new matching or trimming
+     * By default the server execute the routine instantly
+     */
+    void awakeServerThread(
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.hpp:131]: I'd call this function with snake case: `awake_server_thread()`

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.hpp [hunk 3]
```diff
@@ -139,6 +141,11 @@ class PDPServer2 : public aux::PDP
 
 protected:
 
+    /*
+     * Get Pointer to the server resource event thread.
+     */
+    eprosima::fastrtps::rtps::ResourceEvent& getResouceEventThread();
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.hpp:147]: Same here

File: src/cpp/rtps/builtin/discovery/participant/PDPServerListener2.cpp [hunk 4]
```diff
@@ -130,6 +130,8 @@ void PDPServerListener2::onNewCacheChangeAdded(
                 // Take change ownership away from the unique pointer, so that its destruction does not destroy the data
                 pdp_history->remove_change(pdp_history->find_change(change.release()), false);
 
+                // Set timer to 0 ms
+
                 // Ensure processing time for the cache by triggering the Server thread (which process the updates
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServerListener2.cpp]: Merge this two comments

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 5]
```diff
@@ -679,19 +683,34 @@ bool PDPServer2::process_data_queue()
...
+    {
+        logInfo(RTPS_PDP_SERVER, "");
+        logInfo(RTPS_PDP_SERVER, "-------------------- Server routine start --------------------");
+        process_writers_acknowledgements();               // server + ddb(functor_with_ddb)
+        process_data_queue();                             // all ddb
+        process_disposals();                              // server + ddb(changes_to_dispose(), clear_changes_to_disposes())
+        process_dirty_topics();                           // all ddb
+        process_to_send_lists();                          // server + ddb(get_to_send, remove_to_send_this)
+        process_changes_release();                        // server + ddb(changes_to_release(), clear_changes_to_release())
+        result = pending_ack();                           // all server
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:708]: I think we need to explain why `result` comes from this function and what does it mean. Probably we can rename `result` to `pending_work` to make it clearer

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 6]
```diff
@@ -679,19 +683,34 @@ bool PDPServer2::process_data_queue()
...
+        logInfo(RTPS_PDP_SERVER, "-------------------- Server routine end --------------------");
+        logInfo(RTPS_PDP_SERVER, "");
+    }
+    while (!discovery_db_.data_queue_empty());
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:712]: Explain why we need the `do-while`

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 7]
```diff
@@ -1091,6 +1110,11 @@ bool PDPServer2::pending_ack()
     return ret;
 }
 
+eprosima::fastrtps::rtps::ResourceEvent& PDPServer2::getResouceEventThread()
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:1113]: snake case?

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 8]
```diff
@@ -79,13 +79,17 @@ bool PDPServer2::init(
...
         Given the fact that a participant is either a client or a server the
         discoveryServer_client_syncperiod parameter has a context defined meaning.
      */
     mp_sync = new DServerEvent2(this,
-                    TimeConv::Duration_t2MilliSecondsDouble(m_discovery.discovery_config.
-                    discoveryServer_client_syncperiod));
+                    TimeConv::Duration_t2MilliSecondsDouble(
+                            m_discovery.discovery_config.discoveryServer_client_syncperiod));
+
     awakeServerThread();
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:93]: I'd leave the first awaken for the first data received/added to the database, either by listeners or `PDPServer2::announceParticipantState()`

**Review by EduPonz — CHANGES_REQUESTED** [review]
Please rebase to testing

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2764/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/131/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2543/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2776/)

**Review by jparisu — APPROVED** [review]
LGTM

**Pull request merged**